tests/test_autodoc.py

```diff
@@ -1411,22 +1417,248 @@ def test_partialmethod(app):
+
+
+@pytest.mark.sphinx('html', testroot='ext-autodoc')
+def test_autodoc_default_flags__as_dict__no_conversion(app):
+    orig = {
+        'members': 'this,that,other',
+        'undoc-members': None,
+    }
+    app.config.autodoc_default_flags = orig
+    convert_autodoc_default_flags(app, app.config)
+    assert app.config.autodoc_default_flags == orig
+
+
+def diff_no_blanks(a, b):
+    """Produce a line-by-line diff of `a` and `b`, prefixed with "+" or "-".
+
+    This is intended to be used with the output of `do_autodoc`.
+
+    :type a, b: list[str]
+    :rtype: list[str]
+    """
+    # Dictionary iter order is not deterministic
+    ignore = ['dict_proxy', 'mappingproxy']
+    filter_ = lambda line: line.strip() and not any(ig in line for ig in ignore)
+
+    diff = [
+        ('+ '+ line) for line in filter(filter_, b)
+        if line not in a
+    ] + [
+        ('- '+ line) for line in filter(filter_, a)
+        if line not in b
+    ]
+    return diff
+
+
 @pytest.mark.sphinx('html', testroot='ext-autodoc')
 def test_autodoc_default_flags(app):
     # no settings
     actual = do_autodoc(app, 'class', 'target.EnumCls')
     assert '   .. py:attribute:: EnumCls.val1' not in actual
     assert '   .. py:attribute:: EnumCls.val4' not in actual
+    actual = do_autodoc(app, 'class', 'target.CustomIter')
+    assert '   .. py:method:: target.CustomIter' not in actual
 
     # with :members:
-    app.config.autodoc_default_flags = ['members']
+    app.config.autodoc_default_flags = {'members': None}
     actual = do_autodoc(app, 'class', 'target.EnumCls')
     assert '   .. py:attribute:: EnumCls.val1' in actual
     assert '   .. py:attribute:: EnumCls.val4' not in actual
 
     # with :members: and :undoc-members:
-    app.config.autodoc_default_flags = ['members',
-                                        'undoc-members']
+    app.config.autodoc_default_flags = {
+        'members': None,
+        'undoc-members': None,
+    }
     actual = do_autodoc(app, 'class', 'target.EnumCls')
     assert '   .. py:attribute:: EnumCls.val1' in actual
     assert '   .. py:attribute:: EnumCls.val4' in actual
+
+    # with :special-members:
+    # Note that :members: must be *on* for :special-members: to work.
+    # TODO Is this a bug?
+    app.config.autodoc_default_flags = {
+        'members': None,
+        'special-members': None
+    }
+    actual = do_autodoc(app, 'class', 'target.CustomIter')
+    assert '   .. py:method:: CustomIter.__init__()' in actual
+    assert '      Create a new `CustomIter`.' in actual
+    assert '   .. py:method:: CustomIter.__iter__()' in actual
+    assert '      Iterate squares of each value.' in actual
+    if not IS_PYPY:
+        assert '   .. py:attribute:: CustomIter.__weakref__' in actual
+        assert '      list of weak references to the object (if defined)' in actual
+
+
+@pytest.mark.sphinx('html', testroot='ext-autodoc')
+@pytest.mark.parametrize('target, options, expected_diff', [
```

Comment:
I feel this parameterized test is hard to understand...
Could you change this to simple testcases?

In addition, is doing diff important for this case? We've already done testing for each options. So I think the main purpose of this case is `autodoc_default_flags` modifies the options of each autodoc results.

Reply:
Done. I found the diffs useful for myself as I wasn't completely familiar with what each option did but, as you say, there are tests for this already. I've left the tests quite verbose but they don't use `parametrize` or look at the diffs.